File: ab1.cpp
#include <bits/stdc++.h>

using namespace std;
const string BC[] = {".", "O", "X"};
const int value[] = {1000, 1, 11, 121, 1331};
const int mod = 1e9 + 7;
const int len = 4;
const int depth_limit = 10000;
inline int change_player(int n) {
	return n ^ 3;
}
inline int people(int n) {
	return (n == 2) ? -1 : (n == 1 ? 1 : 0);
}
inline void ERROR(void) {
	cout << "\nERROR\n\n\n";
	return;
}

struct TTT {
	int Board[len][len] = {0};
	int filled = 0;
	TTT() {
		for (auto &x : Board)
			for (int &b : x) b = 0;
		filled = 0;
		return;
	}
	int Winner() {
		int winner;
		for (int y = 0; y < len; y++) {
			winner = Board[0][y];
			for (int x = 1; x < len && winner; x++) {
				if (winner != Board[x][y]) winner = 0;
			}
			if (winner) return winner;
		}
		for (int x = 0; x < len; x++) {
			winner = Board[x][0];
			for (int y = 1; y < len && winner; y++) {
				if (winner != Board[x][y]) winner = 0;
			}
			if (winner) return winner;
		}
		winner = Board[0][0];
		for (int s = 1; s < len && winner; s++) {
			if (winner != Board[s][s]) winner = 0;
		}
		if (winner) return winner;
		winner = Board[len - 1][0];
		for (int s = 1; s < len; s++) {
			if (winner != Board[len - 1 - s][s]) winner = 0;
		}
		if (winner) return winner;
		if (filled == len * len) return 0;
		return -1;
	}
	bool put(int x, int y, int player) {
		if (x >= len || y >= len || x < 0 || y < 0 || player < 1 || player > 2) return false;
		if (Board[x][y] != 0) return false;
		Board[x][y] = player;
		filled++;
		return true;
	}
	void clear(void) {
		for (auto &row : Board)
			for (int &i : row) i = 0;
		filled = 0;
		return;
	}
	bool pick(int x, int y, int player) {
		if (x >= len || y >= len || x < 0 || y < 0 || player < 1 || player > 2) return false;
		if (Board[x][y] != player) return false;
		Board[x][y] = 0;
		filled--;
		return true;
	}
};
ostream &operator<<(ostream &s, const TTT &T) {
	string str = "";
	for (int y = 0; y < len; y++) {
		for (int x = 0; x < len; x++) {
			str += BC[T.Board[x][y]];
		}
		str += "\n";
	}
	s << str;
	return s;
}
istream &operator>>(istream &s, TTT &T) {
	const int len2 = len * len;
	char w[len2];
	for (char &i : w) s >> i;
	for (int p = 0; p < len2; p++) {
		if (w[p] == 'O') T.put(p % len, p / len, 1);
		else if (w[p] == 'X') T.put(p % len, p / len, 2);
		else if (w[p] == '.')
			;
		else {
			ERROR();
			return s;
		}
	}
	return s;
}
inline int evaluate(TTT T) {
	int score = 0, winner = T.Winner();
	if (winner >= 0) return value[0] * people(winner);
	else if (winner == 0) return 0;
	for (int count[2] = {0, 0}, y = 0; y < len; y++, count[0] = count[1] = 0) {
		for (int x = 0; x < len; x++) {
			if (T.Board[x][y]) count[T.Board[x][y] & 1]++;
		}
		if ((count[0] == 0 || count[1] == 0) && count[0] + count[1])
			score += value[max(count[0], count[1])] * (count[0] ? -1 : 1);
	}
	for (int count[2] = {0, 0}, x = 0; x < len; x++, count[0] = count[1] = 0) {
		for (int y = 0; y < len; y++) {
			if (T.Board[x][y]) count[T.Board[x][y] & 1]++;
		}
		if ((count[0] == 0 || count[1] == 0) && count[0] + count[1])
			score += value[max(count[0], count[1])] * (count[0] ? -1 : 1);
	}
	int count[2] = {0, 0};
	if (T.Board[0][0]) count[T.Board[0][0] & 1]++;
	if (T.Board[1][1]) count[T.Board[1][1] & 1]++;
	if (T.Board[2][2]) count[T.Board[2][2] & 1]++;
	if ((count[0] == 0 || count[1] == 0) && count[0] + count[1])
		score += value[max(count[0], count[1])] * (count[0] ? -1 : 1);

	count[0] = count[1] = 0;
	if (T.Board[0][2]) count[T.Board[0][2] & 1]++;
	if (T.Board[1][1]) count[T.Board[1][1] & 1]++;
	if (T.Board[2][0]) count[T.Board[2][0] & 1]++;
	if ((count[0] == 0 || count[1] == 0) && count[0] + count[1])
		score += value[max(count[0], count[1])] * (count[0] ? -1 : 1);
	if (score > value[0]) score = value[0];

	return score;
}

pair<int, int> AB(TTT &T, int player, int depth = depth_limit, int a = -mod, int b = mod) {
	int winner = T.Winner();
	if (winner >= 0)	  // game end;
		return {-1, people(winner) * people(player) * value[0]};
	else if (depth == 1)  // lowest depth
		return {-1, evaluate(T) * people(player)};
	int bx = -1, by = -1, score = 0;
	for (int y = 0; y < len; y++)
		for (int x = 0; x < len; x++) {
			if (T.Board[x][y]) continue;
			T.put(x, y, player);
			score = -AB(T, change_player(player), depth - 1, -b, -a).second;
			if (T.pick(x, y, player))
				;
			if (score >= b) return {-1, b};
			if (score > a) a = score, bx = x, by = y;
		}
	if (bx == -1) return {-1, a};
	return {bx + by * len, a};
}

void play() {
	TTT TicTacToe;
	int mode, move, mx, my, len2 = len * len;
	cout << "please input [0:player&player,1:computer&player,2:player&computer] ";
	cin >> mode;
	for (int x, y, player = 1; TicTacToe.filled < len2; player = change_player(player)) {
		cout << TicTacToe << endl;
		if (TicTacToe.Winner() >= 0) {
			cout << "game end" << endl;
			break;
		}
		cout << player << endl;
		if (player & mode) {
			move = AB(TicTacToe, player).first;
			if (move == -1) cout << "ERROR\n";
			if (!TicTacToe.put(move % len, move / len, player)) cout << "ERROR" << endl;
		} else {
			cout << "player " << BC[player] << " input" << endl;
			cin >> mx >> my;
			if (mx == -1 && my == -1) return;
			while (!TicTacToe.put(mx, my, player)) {
				cout << "ERROR please input again" << endl;
				cin >> mx >> my;
			}
		}
	}
	cout << "============================\n";
	cout << TicTacToe;
	cout << "Winner" << TicTacToe.Winner() << endl;
}

File: ab1.py
import copy
char = ['.', 'O', 'X']
value = [100, 2, 2]


class TicTacToe:
    Broad = [[0, 0, 0], [0, 0, 0], [0, 0, 0]]
    Put = 0

    def __init__(self):
        for i in self.Broad:
            for j in i:
                j = 0
        self.Put = 0

    def Winner(self) -> int:
        for i in range(3):
            if (self.Broad[i][0] == self.Broad[i][1] == self.Broad[i][2]):
                return self.Broad[i][0]
            if (self.Broad[0][i] == self.Broad[1][i] == self.Broad[2][i]):
                return self.Broad[0][i]
        if (self.Broad[0][0] == self.Broad[1][1] == self.Broad[2][2] or self.Broad[2][0] == self.Broad[1][1] == self.Broad[0][2]):
            return self.Broad[1][1]
        if (self.Put == 9):
            return -1
        return 0

    def put(self, x: int, y: int, player: int) -> bool:
        if (self.Broad[x][y] != 0):
            return False
        self.Broad[x][y] = player
        self.Put += 1
        return True

    def pick(self, x: int, y: int, player: int) -> bool:
        if (self.Broad[x][y] != player):
            return False
        self.Broad[x][y] = 0
        self.Put -= 1

    def __str__(self) -> str:
        outputstring = ""
        for y in range(3):
            for x in range(3):
                outputstring += char[self.Broad[x][y]]
            outputstring += '\n'
        return outputstring

    def evaluate(self) -> int:
        score: int = 0
        Winner = self.Winner()
        if (Winner > 0):
            return 100 if Winner == 1 else -100
        for i in range(2):
            for j in range(3):
                if (self.Broad[j][i] == self.Broad[j][i+1] != 0 and self.Broad[j][2-2*(i % 2)] == 0):
                    score += (1 if self.Broad[i][j] == 1 else -1)*value[1]
                if (self.Broad[i][j] == self.Broad[i+1][j] != 0 and self.Broad[2-2*(i % 2)][j] == 0):
                    score += (1 if self.Broad[i][j] == 1 else -1)*value[1]
            if (self.Broad[i*2][0] == self.Broad[1][1] != 0 and self.Broad[2-2*(i % 2)][2] == 0):
                score += (1 if self.Broad[i][j] == 1 else -1)*value[1]
            if (self.Broad[i*2][2] == self.Broad[1][1] != 0 and self.Broad[2][2-2*(i % 2)] == 0):
                score += (1 if self.Broad[i][j] == 1 else -1)*value[1]
        for i in range(3):
            if (self.Broad[i][0] == self.Broad[i][2] != 0 and self.Broad[i][1] == 0):
                score += (1 if self.Broad[i][0] == 1 else -1)*value[2]
            if (self.Broad[0][i] == self.Broad[2][i] != 0 and self.Broad[1][i] == 0):
                score += (1 if self.Broad[1][i] == 1 else -1)*value[2]
        if (self.Broad[0][0] == self.Broad[2][2] != 0 and self.Broad[1][1] == 0):
            score += (1 if self.Broad[0][0] == 1 else -1)*value[2]
        if (self.Broad[0][2] == self.Broad[2][0] != 0 and self.Broad[1][1] == 0):
            score += (1 if self.Broad[0][2] == 1 else -1)*value[2]

        return score


def AlphaBeta(G: TicTacToe, a: int, b: int):
    Copy = copy.deepcopy(G)
    player = Copy.Put % 2+1
    s = -1
    winner = G.Winner()
    first = True
    if (winner == player):
        return [100, -1]
    elif (winner == player % 2+1):
        return [-100, -1]
    if (G.Put == 9):
        return [G.evaluate(), -1]
    for y in range(3):
        for x in range(3):
            if (Copy.Broad[x][y] != 0):
                continue
            Copy.put(x, y, player)
            temp = AlphaBeta(copy.deepcopy(Copy), int(-b), int(-a))
            temp[0] = -temp[0]
            Copy.pick(x, y, player)
            if (temp[0] > a):
                a = temp[0]
                s = x+y*3
            if (temp[0] >= b):
                b = temp[0]
                if (b > a):
                    return [temp[0], x+y*3]

    # print(f"{Copy}{a} {s} return a\n")
    return [a, s]


def AB(G: TicTacToe, depth: int, a: int, b: int, isMaxPlayer: bool):
    if (G.Put == 9):
        w = G.Winner()
        if (w == 1 and isMaxPlayer):
            return [100, -1]
        elif (w == 1):
            return [-100, -1]
        elif (w == 2 and isMaxPlayer):
            return [-100, -1]
        else:
            return [100, -1]
    if (depth == 0):
        return [G.evaluate(), -1]
    if (isMaxPlayer == True):
        for x in range(3):
            for y in range(3):
                if (G.Broad[x][y] != 0):
                    continue
                New = copy.deepcopy(G)
                New.put(x, y, 2)
                temp = AB(New, depth-1, int(a), int(b), False)
                if (temp[0] > a):
                    s = x+y*3
                    a = temp[0]
                if (a > b):
                    return [int(a), x+y*3]
        return [int(a), s]
    else:
        for x in range(3):
            for y in range(3):
                if (G.Broad[x][y] != 0):
                    continue
                New = copy.deepcopy(G)
                New.put(x, y, 1)
                temp = AB(New, depth-1, int(a), int(b), True)
                if (temp[0] < b):
                    s = x+y*3
                    b = temp[0]
                if (a > b):
                    return [int(b), x+y*3]
        return [int(b), s]


def Play_computerSecond():
    G = TicTacToe()
    while (G.Winner() == 0):
        while True:
            print(G)
            data = input("請輸入要放入位置")
            data = data.split(" ")
            if (len(data) == 2):
                if (data[0].isdigit() and data[1].isdigit()):
                    if (0 <= int(data[0]) < 3 and 0 <= int(data[1]) < 3):
                        successInput = True
                        G.put(int(data[0]), int(data[1]), 1)
                        break
            print("輸入錯誤")
        print(G)
        if (G.Winner() != 0):
            break
        temp = AlphaBeta(copy.deepcopy(G), -10000, 10000)
        print(G)
        G.put(temp[1] % 3, temp[1]//3, 2)
        print(f"電腦放({temp[1]%3},{temp[1]//3})  {temp[1]}")
        print(G)


Play_computerSecond()
